pkg/cmd/roachtest/decommission.go

```diff
@@ -374,318 +451,486 @@ func runDecommissionAcceptance(ctx context.Context, t *test, c *cluster) {
 	{
-		o, err := decommission(ctx, 3, c.Node(2),
-			"decommission", "--wait=all", "--format=csv")
-		if err != nil {
-			t.Fatalf("decommission failed: %v", err)
+		targetNodeA, targetNodeB := decommissionedNodeA, decommissionedNodeB
+		if targetNodeB < targetNodeA {
+			targetNodeB, targetNodeA = targetNodeA, targetNodeB
 		}
 
-		exp := [][]string{
-			decommissionHeader,
-			{"2", "true", "0", "true", "false"},
-			decommissionFooter,
+		runNode := getRandNode()
+		waitStrategy := "all" // Blocking decommission.
+		if i := rand.Intn(2); i == 0 {
+			waitStrategy = "none" // Polling decommission.
 		}
-		if err := matchCSV(o, exp); err != nil {
-			t.Fatal(err)
-		}
-	}
 
-	t.l.Printf("recommissioning second node from itself\n")
-	if _, err := decommission(ctx, 2, c.Node(2), "recommission"); err != nil {
-		t.Fatalf("recommission failed: %v", err)
-	}
+		t.l.Printf("fully decommissioning [n%d,n%d] from n%d, using --wait=%s\n",
+			targetNodeA, targetNodeB, runNode, waitStrategy)
 
-	t.l.Printf("decommissioning third node (from itself)\n")
-	func() {
-		// This should not take longer than five minutes, and if it does, it's
-		// likely stuck forever and we want to see the output.
-		timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
-		defer cancel()
-		o, err := decommission(timeoutCtx, 3, c.Node(3),
-			"decommission", "--wait=all", "--format=csv")
-		if err != nil {
-			t.Fatalf("decommission failed: %v", err)
+		// When using --wait=none, we poll the decommission status.
+		maxAttempts := 50
+		if waitStrategy == "all" {
+			// --wait=all is a one shot attempt at decommissioning, that polls
+			// internally.
+			maxAttempts = 1
 		}
 
-		exp := [][]string{
-			decommissionHeader,
-			{"3", "true", "0", "true", "false"},
-			decommissionFooter,
-		}
-		if err := matchCSV(o, exp); err != nil {
+		if err := retry.WithMaxAttempts(ctx, retryOpts, maxAttempts, func() error {
+			o, err := decommission(ctx, c.Nodes(targetNodeA, targetNodeB), runNode,
+				fmt.Sprintf("--wait=%s", waitStrategy), "--format=csv")
+			if err != nil {
+				t.Fatalf("decommission failed: %v", err)
+			}
+
+			exp := [][]string{
+				decommissionHeader,
+				{strconv.Itoa(targetNodeA), "true", "0", "true", "decommissioned", "false"},
+				{strconv.Itoa(targetNodeB), "true", "0", "true", "decommissioned", "false"},
+				decommissionFooter,
+			}
+			return matchCSV(o, exp)
+		}); err != nil {
 			t.Fatal(err)
 		}
-	}()
 
-	// Now that the third node is down and decommissioned, decommissioning it
-	// again should be a no-op. We do it from node one but as always it doesn't
-	// matter.
-	t.l.Printf("checking that other nodes see node three as successfully decommissioned\n")
-	{
-		o, err := decommission(ctx, 2, c.Node(3),
-			"decommission", "--format=csv") // wait=all is implied
-		if err != nil {
-			t.Fatalf("decommission failed: %v", err)
+		// Check that even though the node is decommissioned, we still see it (since
```

Comment:
even though two nodes are decommissioned, we still see them (since they remain live)